### ISSUE TITLE
Allow birthYear as an option field for users

### ISSUE THREAD

**Issue by astroash** [TwinePlatform/twine-api#289]
- [x] creating visitors should take birthYear as an optional field
- [x] visitor filters should take ~~'prefer not to say'~~ `null` as an age range 
- [ ] aggregated data should return data for ~~'prefer not to say'~~ `null` age range 

TwinePlatform/hq#40
TwinePlatform/twine-visitor#584

**Comment by astroash** [comment]
Entering 'prefer not to say' as `null` in the database for birthYear will then require us to map all user queries to enter this value.

Currently user data is being pulled from 6+ model methods. As this grows I can easily see this being forgotten about and returning inaccurate data to clients. I am keen to change the birthYear column type from `int` to `text` in the database to allow us to enter 'prefer not to say'.

@eliascodes thoughts?

**Comment by eliascodes** [comment]
Not massively keen on this as I think it conflates a presentational issue with a data representation issue.

`"Prefer not to say"` is just what we want the user to see instead of `NULL`. It's the clients who should be dealing with basically all presentational logic. A "null" value in `birthYear` should be interpreted as `prefer not to say`, where appropriate, by the client.

**Comment by eliascodes** [comment]
Actually, my bad, another place this could be handled would be in the `User.serialise` method. We'd need to double check the handlers to make sure they're actually using this method before sending responses though.

I'm not opposed to the above because the `serialise` method is specifically meant for transforming models to being consumable by (all) clients.

> should setting the age value to null be the responsibility of the handler or the User.add model method. Currently for optional gender field this is handled by a joi default value.

Which handler are we talking about here? The visitor register handler doesn't set a default value, that's taken care of by the data model (as it should be).

> should our optional gender value be dealt with in the same way - eg saved in the db as null and displayed as 'prefer not to say' by the client

I think since gender/disability/ethnicity already work, I'm not that keen on changing them. 

<details>
<summary>TLDR</summary>
<p>
Besides, IMO gender (and disability/ethnicity) is a bit different. The problem is encoding a non-answer. For gender, the possible answers are fixed, enumerable, few and text-based. It's simple to encode a "non-answer" as simply an extra option. For age, the possible answers are not fixed, few or text-based (though they are enumerable). So the options are choose a less natural data type (text instead of int), or use `NULL`. The former means I'm unable to do arithmetic directly with the data, and I have no easy way of enforcing consistency on the "non-answer" option (e.g. I'll need to write application code to ensure only `"prefer not to say"` is allowed, as opposed to `"Prefer not to say"` or `"no answer"` or whatever). The latter means extra presentational logic, placed at some point in the stack. It's an imperfect trade-off.
<p>
</details>

**Comment by astroash** [comment]
> Which handler are we talking about here?

https://github.com/TwinePlatform/twine-api/blob/06f514bc44e6a1bde7d0c7a75c1f44871ec07a3c/src/api/v1/users/register/volunteers.ts#L41

`register/volunteers`, gender was not optional in the visitor app when we wrote that handler. Now that we are trying to make optional user fields consistent across both apps I think we should decide where we are adding default values.

> this could be handled would be in the User.serialise method

I've come round to this being a client side responsibility, but also happy for this approach. The only down side for this how we display the above mentioned aggregated data. We will either need a `VisitLogs.serialise` method or accepting that we're naming that value _'prefer not to say'_ in the model method.

> TLDR

I like the idea of `null` being used for data a user doesn't want to answer. _'prefer not to say'_ is not a recognised gender or disability, it's just how we have decided to phase the opt out. In the future it say become more 'politically correct' to describe those who don't want to share their gender as 'genderless' or 'human', by separating out the phrasing of opting out it will be easier in the future. _Also, personally I don't like our gender options as they're not inclusive, you shouldn't have binary genders or keep it to yourself, so any ability to easily change this in the future IMO is better._

If we do go for _'prefer not to say'_ being added by the client then all non specified genders/disabilities/ethnicities/ages can be sent in the payload as `null` (or not send and added in the handler/methods.)

_PS side note, [you're using TLDR incorrectly](https://www.lifewire.com/what-is-tldr-2483633)_

**Comment by eliascodes** [comment]
The gender/ethnicity/disability values are foreign keys into other tables. Having null values means we need to change the relevant `innerJoin`s to `outerJoin`s; making a mistake or missing one risks returning erroneous data.

Any changes to gender/ethnicity/disability can be implemented separately. Can we just do the birth year changes for now, since it's actually a feature that was asked for. Let's start a new issue for the other stuff.

**Comment by astroash** [comment]
Sure. Do you have any preference for my other comments?

**Comment by astroash** [comment]
Cool, will look into it. 

Also, our app still caches, there's just no broken service worker. If we want it not to cache we need to do [this](https://stackoverflow.com/questions/49604821/cache-busting-with-create-react-app). Tho, I don't think we should avoid changes due to caching.

**Comment by eliascodes** [comment]
> app still caches

Shouldn't do, I explicitly unregister the service worker:
https://github.com/TwinePlatform/twine-visitor/blob/1d2a49be79cc86e3c4bfb69c41327d1d034fff31/src/index.js#L17

**Comment by astroash** [comment]
From my understanding service workers and caching are two separate things. The browser can still cache the js/css/html files even if we have unregistered the service worker. Link above is to a SO answer about cache busting with CRA.

**Comment by eliascodes** [comment]
> service workers and caching are two separate things

They are, I was being imprecise.
